lib/Command/RecreateMasterKey.php

```diff
@@ -0,0 +1,338 @@
+
+	/** @var Setup  */
+	protected $userSetup;
+
+	/** @var IMailer  */
+	protected $mailer;
+
+	/** @var ISecureRandom  */
+	protected $secureRandom;
+
+	/** @var IL10N  */
+	protected $l;
+
+	/** @var ILogger  */
+	protected $logger;
+
+	/** @var  */
+	protected $encryptAll;
+
+	/** @var array files which couldn't be decrypted */
+	protected $failed;
+
+	/**
+	 * RecreateMasterKey constructor.
+	 *
+	 * @param IUserManager $userManager
+	 * @param View $rootView
+	 * @param KeyManager $keyManager
+	 * @param Util $util
+	 * @param IAppManager $IAppManager
+	 * @param IAppConfig $appConfig
+	 * @param IConfig $config
+	 * @param ISession $session
+	 * @param QuestionHelper $questionHelper
+	 * @param Setup $userSetup
+	 * @param IMailer $mailer
+	 * @param ISecureRandom $secureRandom
+	 * @param IL10N $l
+	 * @param ILogger $logger
+	 */
+	public function __construct(IUserManager $userManager, View $rootView, KeyManager $keyManager, Util $util,
+								IAppManager $IAppManager, IAppConfig $appConfig, IConfig $config, ISession $session,
+								QuestionHelper $questionHelper, Setup $userSetup, IMailer $mailer,
+								ISecureRandom $secureRandom, IL10N $l, ILogger $logger) {
+
+		parent::__construct();
+		$this->userManager = $userManager;
+		$this->rootView = $rootView;
+		$this->keyManager = $keyManager;
+		$this->util = $util;
+		$this->IAppManager = $IAppManager;
+		$this->appConfig = $appConfig;
+		$this->config = $config;
+		$this->session = $session;
+		$this->questionHelper = $questionHelper;
+		$this->userSetup = $userSetup;
+		$this->mailer = $mailer;
+		$this->secureRandom = $secureRandom;
+		$this->l = $l;
+		$this->logger = $logger;
+	}
+
+	protected function configure() {
+		parent::configure();
+
+		$this
+			->setName('encryption:recreate-master-key')
+			->setDescription('Replace existing master key with new one. Encrypt the file system with newly created master key')
+		;
+
+		$this->addOption(
+			'yes',
+			'y',
+			InputOption::VALUE_NONE,
+			'Answer yes to all questions'
+		);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$yes = $input->getOption('yes');
+		if ($this->util->isMasterKeyEnabled()) {
+			$question = new ConfirmationQuestion(
+				'Warning: Inorder to re-create master key, the entire ownCloud filesystem will be decrypted and then encrypted using new master key.'
+				. ' Do you want to continue? (y/n)', false);
+			if ($yes || $this->questionHelper->ask($input, $output, $question)) {
+
+				$output->writeln("Decryption started\n");
+				$progress = new ProgressBar($output);
+				$progress->start();
+				$progress->setMessage("Decryption progress...");
+				$progress->advance();
+
+				$this->decryptAllUsersFiles($progress, $output);
+				$progress->finish();
+
+				if (empty($this->failed)) {
+
+					$this->IAppManager->disableApp('encryption');
+
+					//Delete the files_encryption dir
+					$this->rootView->deleteAll('files_encryption');
```

Comment:
is this correct ? Remember that encryption keys can be stored in a different location with an occ storage (key storage root)